### PR TITLE
Fix shebang so brew bundle works on Linux

### PR DESCRIPTION
**Pull request by adamliter**
See the discussion in https://github.com/Homebrew/homebrew-bundle/issues/1.

This pull request fixes the shebang in `/cmd/brew-bundle` so that it can run on Linux. In Linux, the `#!/usr/bin/env` shebang previously tried to run the executable "ruby -W0", which returned the error:

    /usr/bin/env: ruby -W0: No such file or directory

(See, for example, [Bash trace mode (bash -x) in the shebang](http://stackoverflow.com/q/24032641/2571049) on Stack Overflow.)

By changing this to `#!/bin/sh` and making the call to ruby with `exec ruby -x "$0" "$@"`, this problem is solved.

`man ruby` tells us for the flag `-x` that:

>      -x[directory]  Tells Ruby that the script is embedded in a message.  Leading garbage will be discarded until the first line that starts with ``#!'' and con-
                    tains the string, ``ruby''.  Any meaningful switches on that line will be applied.  The end of the script must be specified with either EOF,
                    ^D (control-D), ^Z (control-Z), or the reserved word __END__.  If the directory name is specified, Ruby will switch to that directory before
                    executing script.

This is why `/cmd/brew-bundle` now has `__END__` at the end of the file.

I've tested this on both Mac (10.10.5) and Ubuntu (14.04).

### CODE REVIEW

File: cmd/brew-bundle [hunk 1]
```diff
@@ -1,4 +1,6 @@
-#!/usr/bin/env ruby -W0
+#!/bin/sh
```
Comment (mikemcquaid) [inline cmd/brew-bundle]: Let's just try `/usr/bin/env ruby` first, thanks.

**Comment by adamliter** [comment]
@mikemcquaid Okay, I've changed the commit to just do `#!/usr/bin/env ruby` as the shebang, and I've tested it on Ubuntu 14.04. It works.

**Comment by mikemcquaid** [comment]
@xu-cheng Any objections here?

**Comment by xu-cheng** [comment]
Maybe add `$VERBOSE = nil` which is the equivalence of `-W0`

**Comment by mikemcquaid** [comment]
:+1:. Mind giving that a go @adamliter? Also, can you update the `README` to remove the stuff about this not working on Linux?

**Comment by adamliter** [comment]
@mikemcquaid Yes, I can do that, but it'll probably have to wait a day or two. I'm currently on vacation and in a place with very sparse internet connection. :stuck_out_tongue: Thanks!

**Comment by mikemcquaid** [comment]
No rush!

**Comment by adamliter** [comment]
@mikemcquaid @xu-cheng I made the suggested changes, but now I'm getting:

```sh
Error: Unknown command: cask
Error: Unknown command: cask
0.9.5
```

When trying to run any `brew bundle` command on Linux. I'm not sure why this is showing up now. Let me see if I can figure out where it is coming from. I definitely wouldn't merge this pull request yet, though.

**Comment by mikemcquaid** [comment]
@adamliter thanks!

**Comment by dunn** [comment]
For some reason the verbosity, in particular `Kernel.exit` is leading to this final item on `brew bundle dump`s:
```
[...]
cask 'Kernel.exit'

**Comment by adamliter** [comment]
@dunn Hmm, that's interesting. Any idea why?

I haven't had much of a chance to do any debugging yet, but I just switched back to what was in the original pull request:

```sh
#!/bin/sh
exec ruby -x "$0" "$@"
#!ruby -W0
...
__END__
```

And I'm getting the same messages:

```sh
Error: Unknown command: cask
Error: Unknown command: cask
0.9.5
```

I'm not sure if something has changed or if I just missed that in my original testing (no idea how I would have?).

I'll try both things again with a fresh install of Linuxbrew and Homebrew-bundle sometime later.

It almost seems as if the output from `brew.installed?` and `cask.installed?` in `utils.rb` is getting through, but that wouldn't make sense since it is piped to `/dev/null`. And it also wouldn't explain the different behavior from Mac to Linux.

I'll try to do some closer looking tomorrow, but this might be beyond me to figure out. I'm not really a programmer. :stuck_out_tongue:

**Comment by mikemcquaid** [comment]
Thanks for digging in here @adamliter. We'll make a programmer out of you yet :wink:

**Comment by dunn** [comment]
Still investigating, but dropping these findings in case someone else connects the dots faster:
- `Bundle.cask_installed?` returns true even if Cask isn't installed.
- internal calls to `brew cask` fail even if Cask _is_ installed.

**Comment by dunn** [comment]
@adamliter was on the right track.  Maybe we were using a non-standard redirection idiom, since this fixes it:
```diff
diff --git a/lib/bundle/utils.rb b/lib/bundle/utils.rb
index 8c12fa9..5e6b6a8 100644
--- a/lib/bundle/utils.rb
+++ b/lib/bundle/utils.rb
@@ -17,15 +17,15 @@ module Bundle
   end

   def self.brew_installed?
-    @@brew ||= Kernel.system("brew --version &>/dev/null")
+    @@brew ||= Kernel.system("brew --version >/dev/null 2>&1")
   end

   def self.cask_installed?
-    @@cask ||= Kernel.system("brew cask --version &>/dev/null")
+    @@cask ||= Kernel.system("brew cask --version >/dev/null 2>&1")
   end

   def self.services_installed?
-    @@services ||= Kernel.system("brew services --help &>/dev/null")
+    @@services ||= Kernel.system("brew services --help >/dev/null 2>&1")
   end

   def self.brewfile
```

**Comment by adamliter** [comment]
@dunn Hmm, I wonder why the non-standard redirect idiom worked on Mac but not Linux. Interesting.

But anyway, should I add your changes to the pull request?

**Comment by mikemcquaid** [comment]
It will probably depend on the shell it's calling out to. No harm in adding it.

**Comment by adamliter** [comment]
With the fixes to the redirects proposed by @dunn, it seems like everything is working. I tested all of the `brew bundle` commands on a Linux VM (Ubuntu 14.04).

**Comment by dunn** [comment]
Awesome! Most of the credit goes to you for narrowing in on the problem.

**Comment by DomT4** [comment]
`&>` isn't supported by POSIX `sh`, so if the call is sent to `/bin/sh` rather than `/bin/bash` it can fail.

**Comment by adamliter** [comment]
@DomT4 Thanks for the clarification! So the fix that @dunn suggested and that I added to this pull request would be the better way to go, then?

**Comment by DomT4** [comment]
Yeah, you'll hit less potential issues using the `>/dev/null 2>&1` syntax, which should work well for dash/bash/zsh/fish/sh.

**Comment by DomT4** [comment]
You may have hit this issue on Linux in particular because the Linux distros started to favour `sh` calling `dash` a while ago, and `echo "dog" &>/dev/null` doesn't work as intended with `dash`.

**Comment by adamliter** [comment]
@DomT4 The odd thing is that your suggested test string doesn't print any output on my Linux VM (Ubuntu 14.04), whereas I *was* seeing the gratuitous output from `brew bundle` prior to the fix from @dunn.

**Comment by DomT4** [comment]
So
```bash
sh -c "echo dog &>/dev/null"
```
Doesn't throw anything to stdout on Ubuntu? That's interesting. What does
```bash
sh --version
```
Throw, out of curiosity?

**Comment by adamliter** [comment]
@DomT4 Ah, so `sh -c "echo dog &>/dev/null"` does print "dog" to stdout.

`sh --version` returns `sh: 0: Illegal option --`.

**Comment by DomT4** [comment]
> does print "dog" to stdout.

:+1:. I presume `/bin/sh -c "echo dog >/dev/null 2>&1"` doesn't?

> returns sh: 0: Illegal option --.

Heh, forgot `dash` doesn't support that, or version checks in general. Sorry about that!

**Comment by adamliter** [comment]
> :+1:. I presume sh -c "echo dog >/dev/null 2>&1" doesn't?

Correct. :smile_cat:

**Comment by DomT4** [comment]
Cool. Unless anyone objects, I think adopting Alex's syntax is the path forwards. It's friendlier towards a wider variety of shells than `&>`.

**Pull request merged**